asmcomp/amd64/emit.mlp

```diff
@@ -835,9 +902,49 @@ let fundecl fundecl =
       cfi_adjust_cfa_offset n;
     end;
   end;
-  def_label !tailrec_entry_point;
-  emit_all true fundecl.fun_body;
-  List.iter emit_call_gc !call_gc_sites;
+
+  let preproc_res = preproc_fun fundecl.fun_body fundecl.fun_name in
+  let s = !Clflags.stack_slop + preproc_res.max_stack_size in
+  (* Extra 1k bytes slop space if profiling is enabled. Ensures that `mcount`
+   * and signal handlers do not stack overflow. *)
+  let s = if !Clflags.gprofile then s + 1024 else s in
+  (* Extra 48 bytes for caml_context at the bottom of OCaml stack if the
+   * function makes external calls *)
+  let s = if preproc_res.contains_external_calls then s + 48 else s in
+
+  if not preproc_res.contains_nontail_calls && 
+     s < stack_threshold_size 
+  then begin
+    def_label !tailrec_entry_point;
+    emit_all true fundecl.fun_body;
+    List.iter emit_call_gc !call_gc_sites;
+  end
+  else begin
+    (* XXX KC: fundecl.fun_body.live does not capture liveness at the start of
+    the function properly. *)
```

Comment:
I assume you are referring to the `XXX KC` comment. Earlier during development, I observed that the live register set at the start of a function body did not capture live registers properly. *This is no longer true*. I suspect liveness analysis (ran using `ocamlrun ocamlopt`) gave incorrect results due to some subtle bugs with compaction code that I only fixed later. I do the sensible thing now and it works. See https://github.com/ocamllabs/ocaml-effects/commit/ca68a3b70328991246e134964867e4814dc637ca.